vendor/github.com/kubernetes-csi/csi-lib-utils/accessmodes/access_modes.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package accessmodes
+
+import (
+	"fmt"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	v1 "k8s.io/api/core/v1"
+)
+
+// ToCSIAccessMode maps PersistentVolume access modes in Kubernetes to CSI
+// access modes.
+func ToCSIAccessMode(pvAccessModes []v1.PersistentVolumeAccessMode) (csi.VolumeCapability_AccessMode_Mode, error) {
+	m := uniqueAccessModes(pvAccessModes)
+
+	switch {
+	case m[v1.ReadWriteMany]:
+		// ReadWriteMany takes precedence, regardless of what other
+		// modes are set.
+		return csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER, nil
+
+	case m[v1.ReadOnlyMany] && m[v1.ReadWriteOnce]:
+		// This is not possible in the CSI spec.
+		return csi.VolumeCapability_AccessMode_UNKNOWN, fmt.Errorf("CSI does not support ReadOnlyMany and ReadWriteOnce on the same PersistentVolume")
+
+	case m[v1.ReadOnlyMany] && m[v1.ReadWriteOncePod]:
+		// This is not possible in the CSI spec.
+		return csi.VolumeCapability_AccessMode_UNKNOWN, fmt.Errorf("CSI does not support ReadOnlyMany and ReadWriteOncePod on the same PersistentVolume")
+
+	case m[v1.ReadWriteOnce] && m[v1.ReadWriteOncePod]:
+		// This is not possible in the CSI spec.
+		return csi.VolumeCapability_AccessMode_UNKNOWN, fmt.Errorf("CSI does not support ReadWriteOnce and ReadWriteOncePod on the same PersistentVolume")
+
+	case m[v1.ReadOnlyMany]:
+		return csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY, nil
+
+	case m[v1.ReadWriteOnce]:
+		return csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER, nil
+
+	// This mapping exists to enable CSI drivers that lack the
+	// SINGLE_NODE_MULTI_WRITER capability to work with the
+	// ReadWriteOncePod access mode.
+	case m[v1.ReadWriteOncePod]:
+		return csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER, nil
+
+	default:
+		return csi.VolumeCapability_AccessMode_UNKNOWN, fmt.Errorf("unsupported AccessMode combination: %+v", pvAccessModes)
+	}
+}
+
+// ToSingleNodeMultiWriterCapableCSIAccessMode maps PersistentVolume access
+// modes in Kubernetes to CSI access modes for drivers that support the
+// SINGLE_NODE_MULTI_WRITER capability.
+func ToSingleNodeMultiWriterCapableCSIAccessMode(pvAccessModes []v1.PersistentVolumeAccessMode) (csi.VolumeCapability_AccessMode_Mode, error) {
```

Comment:
Just wondering, would it be better to have one function with an extra argument `ToCSIAccessMode(..., singleNodeMultiWriterSupported bool)`? With two functions, all callers must do the same
```
if singleNodeMultiWriterSupported { 
   am, err = ToSingleNodeMultiWriterCapableCSIAccessMode(...)
} else {
   am, err= ToCSIAccessMode(...)
}
```

Reply:
I think having a boolean argument for this is a bit of a code smell but since we're only dealing with one it's ok. I'll make this change in the library. Also assigned you to the PR for that, https://github.com/kubernetes-csi/csi-lib-utils/pull/89